### PR TITLE
Rework the PDict objects in Session to use std::map

### PR DESCRIPTION
**Pull request by timwoj**
This is the second in a series for PRs for #472. This PR replaces the following instances of `PDict` with `std` versions:

- `Connection` tracking in `Session`, replaced with a `std::map` keyed on a `ConnIDKey` structure
- `FragReassembler` tracking in `Session`, replaced with a `std::map` keyed on a `std::tuple` with the same data that we were hashing previously.

This PR also includes a bunch of changes to the output from unit tests. This is due to the fact that the elements in the `map`s are ordered differently than they would have been in the `PDict` obejcts. When `NetSessions::Drain` is called, it outputs the remainder of the connections into the logs. This causes the output to be different than it was previously. The output is consistent from one run to the next after the changes, since the sorting remains consistent. I spent a good amount of time running diffs between the two outputs to make sure that all of the connections still exist in each of the logs, and that nothing changes except the ordering and the UIDs assigned (since they're assigned during `Drain`).

These changes are probably easiest to review commit-by-commit, since they're fairly self-contained even though they're modifying the same files.

Timing against the `2009-M57-day11-18.trace` file, average of ten passes over the file:

Old: 60.538s
New: 57.677s

Memory usage against the same file, average of ten passes over the file (measured with `gtime -v` and the "Maximum resident set size" value):

Old: 96437.2 bytes
New: 93390.4 bytes

### CODE REVIEW

File: src/Sessions.h [hunk 1]
```diff
@@ -212,12 +216,16 @@ class NetSessions {
 	bool CheckHeaderTrunc(int proto, uint32_t len, uint32_t caplen,
 			      const Packet *pkt, const EncapsulationStack* encap);
 
+	void InsertConnection(ConnectionMap* m, const ConnIDKey& key, Connection* conn);
```
Comment (jsiwek) [inline src/Sessions.h]: Might helpful to document "must not already be in the map" as a pre-condition for this function.

File: src/Conn.h [hunk 2]
```diff
@@ -90,8 +89,9 @@ class Connection : public BroObj {
 			// arguments for reproducing packets
 			const Packet *pkt);
 
-	HashKey* Key() const			{ return key; }
-	void ClearKey()				{ key = 0; }
+	ConnIDKey& Key()				{ return key; }
```
Comment (jsiwek) [inline src/Conn.h]: Can return const-reference?

File: src/Conn.h [hunk 3]
```diff
@@ -90,8 +89,9 @@ class Connection : public BroObj {
...
+	void ClearKey()					{ key_valid = false; }
+	bool IsKeyValid() const			{ return key_valid; }
```
Comment (jsiwek) [inline src/Conn.h]: Might be helpful to document what these functions really mean.  "i.e. we call the key valid only when the connection is contained in the session map".

File: src/Sessions.cc [hunk 4]
```diff
@@ -940,27 +929,22 @@ FragReassembler* NetSessions::NextFragment(double t, const IP_Hdr* ip,
...
 
-	FragReassembler* f = fragments.Lookup(h);
 	if ( ! f )
 		{
-		f = new FragReassembler(this, ip, pkt, h, t);
-		fragments.Insert(h, f);
-		Unref(key);
+		f = new FragReassembler(this, ip, pkt, key, t);
+		fragments[key] = f;
+		if ( fragments.size() > stats.max_fragments )
```
Comment (jsiwek) [inline src/Sessions.cc]: Signed/unsigned comparison warning.  Also similar warnings later on for the TCP/UDP/ICMP stats.  Maybe just bump `SessionStats` counters to `size_t` ?

**Comment by timwoj** [comment]
Addressed all of the comments and rebased in the fixes.

**Pull request merged**